src/test/custom_loader_test.cpp

```diff
@@ -0,0 +1,967 @@
+        std::filesystem::create_directories(cl_model_1_path);
+    }
+    void TearDown() {
+        // Clean up temporary destination
+        std::filesystem::remove_all(cl_models_path);
+    }
+    /**
+     * @brief This function should mimic most closely predict request to check for thread safety
+     */
+    void performPredict(const std::string modelName,
+        const ovms::model_version_t modelVersion,
+        const tensorflow::serving::PredictRequest& request,
+        std::unique_ptr<std::future<void>> waitBeforeGettingModelInstance = nullptr,
+        std::unique_ptr<std::future<void>> waitBeforePerformInference = nullptr);
+
+    void deserialize(const std::vector<float>& input, InferenceEngine::InferRequest& inferRequest, std::shared_ptr<ovms::ModelInstance> modelInstance) {
+        auto blob = InferenceEngine::make_shared_blob<float>(
+            modelInstance->getInputsInfo().at(DUMMY_MODEL_INPUT_NAME)->getTensorDesc(),
+            const_cast<float*>(reinterpret_cast<const float*>(input.data())));
+        inferRequest.SetBlob(DUMMY_MODEL_INPUT_NAME, blob);
+    }
+
+    void serializeAndCheck(int outputSize, InferenceEngine::InferRequest& inferRequest) {
+        std::vector<float> output(outputSize);
+        ASSERT_THAT(output, Each(Eq(0.)));
+        auto blobOutput = inferRequest.GetBlob(DUMMY_MODEL_OUTPUT_NAME);
+        ASSERT_EQ(blobOutput->byteSize(), outputSize * sizeof(float));
+        std::memcpy(output.data(), blobOutput->cbuffer(), outputSize * sizeof(float));
+        EXPECT_THAT(output, Each(Eq(2.)));
+    }
+
+    ovms::Status performInferenceWithRequest(const tensorflow::serving::PredictRequest& request, tensorflow::serving::PredictResponse& response) {
+        std::shared_ptr<ovms::ModelInstance> model;
+        std::unique_ptr<ovms::ModelInstanceUnloadGuard> unload_guard;
+        auto status = ovms::getModelInstance(manager, "dummy", 0, model, unload_guard);
+        if (!status.ok()) {
+            return status;
+        }
+
+        response.Clear();
+        return ovms::inference(*model, &request, &response, unload_guard);
+    }
+
+public:
+    ConstructorEnabledModelManager manager;
+    ovms::ModelConfig config = DUMMY_MODEL_CONFIG;
+    ~TestCustomLoader() {
+        std::cout << "Destructor of TestCustomLoader()" << std::endl;
+    }
+
+    std::string cl_models_path;
+    std::string cl_model_1_path;
+    std::string cl_model_2_path;
+};
+
+::grpc::Status test_PerformModelStatusRequestX(ModelServiceImpl& s, tensorflow::serving::GetModelStatusRequest& req, tensorflow::serving::GetModelStatusResponse& res) {
+    spdlog::info("reqx={} resx={}", req.DebugString(), res.DebugString());
+    ::grpc::Status ret = s.GetModelStatus(nullptr, &req, &res);
+    spdlog::info("returned grpc status: ok={} code={} msg='{}'", ret.ok(), ret.error_code(), ret.error_details());
+    return ret;
+}
+
+void TestCustomLoader::performPredict(const std::string modelName,

```

Comment:
I don't know the C++ lvl API for inference. @atobiszei , can you look if this is all right?

Reply:
This is copy of test code from src/test/prediction_service_test.cpp.

Why we couldn't make it shared code? Copying code (especially this amount) is almost always bad.

Couldn't you move class out of src/test/prediction_service_test.cpp & then use either use inheritance to do custom setUp()? As I think this is the main reason why you couldn't/didn't want to put those tests there

Reply:
The functionality apart from setup is different as we start from file, not from config as done in prediction service. Also did not want to change existing source files to the extent possible and have all the custom loader tests at one place.

Reply:
Will have a relook at all the functions/part of code reused from the existing tests